### PR TITLE
MRPHS-3254: Assign static ip to vmware vm at the time of provisioning

### PR DESCRIPTION
**Pull request by avnish30jn**
**Jira-Ticket**

**Problem**

Assign static ip to vmware vm at the time of provisioning. 

**Solution**

Created a custom spec in vcenter db with name 'static-ip-libretto' if it doesn't exist.
Get the static ip and updated the spec and passed it to the vm during clone.

**Unit testing**

Tried creating vm with the below inputs:
1) static ip and custom spec does not exist in vcenter db
SUCCESS- VM is created successfully with the static ip and custom spec is created in vcenter db.
2) static ip and custom spec exists in vcenter db
SUCCESS- VM is created successfully with the static ip.
3) static ip is not sent
SUCCESS- VM is created successfully with the ip assigned via dhcp.

Logs below:

**No static-ip**
```
input :- {"Host":"dev-vcenter.gsintlab.com","Destination":{"DestinationName":"C3 Cluster-1","DestinationType":"cluster","HostSystem":"67.220.186.4"},"Username":"","Password":"","Insecure":true,"Datacenter":"C3 DataCenter","OvaPathUrl":"apporbit-visor-2.1.ova","Networks":[{"name":"VM Private"}],"Name":"avni_2","Size":"small","Datastores":["datastore2-ESXi13-2TB"],"Template":"visor_template_2.1","SkipExisting":2}


[root@my_machine test]# go run vsphereclient.go
Ip List : {"IPs":["10.10.24.120"],"Id":"vm-5019"}
[root@my_machine test]#
```

**with static-ip**

```
input:- {"ipsetting":{"ip":"10.10.24.239","subnet_mask":"255.255.255.0","gateway":""},"Host":"dev-vcenter.gsintlab.com","Destination":{"DestinationName":"C3 Cluster-1","DestinationType":"cluster","HostSystem":"67.220.186.4"},"Username":"","Password":"","Insecure":true,"Datacenter":"C3 DataCenter","OvaPathUrl":"apporbit-visor-2.1.ova","Networks":[{"name":"VM Private"}],"Name":"avni_2","Size":"small","Datastores":["datastore2-ESXi13-2TB"],"Template":"visor_template_2.1","SkipExisting":2}


[root@my_machine test]# go run vsphereclient.go
Ip List : {"IPs":["10.10.24.239"],"Id":"vm-5024"}
[root@my_machine test]#
```

### CODE REVIEW

File: virtualmachine/vsphere/util.go [hunk 1]
```diff
@@ -585,12 +586,41 @@ var cloneFromTemplate = func(vm *VM, dcMo *mo.Datacenter, usableDatastores []str
...
+	exists, err := customizationSpecManager.DoesCustomizationSpecExist(
+		vm.ctx, STATICIP_CUSTOM_SPEC_NAME)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		err = createCustomSpecStaticIp(vm)
+		if err != nil {
+			return fmt.Errorf("Error creating custom spec: %s",
```
Comment (ashpuri) [inline virtualmachine/vsphere/util.go]: errors should be printed with %v  -- same comment applicable below

File: virtualmachine/vsphere/util.go [hunk 2]
```diff
@@ -585,12 +586,41 @@ var cloneFromTemplate = func(vm *VM, dcMo *mo.Datacenter, usableDatastores []str
...
+				err)
```
Comment (ashpuri) [inline virtualmachine/vsphere/util.go]: move it to line above -- same comment applicable below

File: virtualmachine/virtualmachine.go [hunk 3]
```diff
@@ -27,6 +27,12 @@ type VirtualMachine interface {
 	RemoveDisk([]string) error
 }
 
+type IpSetting struct {
```
Comment (ashpuri) [inline virtualmachine/virtualmachine.go]: IpSetting --> NetworkSettings?

File: virtualmachine/vsphere/util.go [hunk 4]
```diff
@@ -585,12 +586,41 @@ var cloneFromTemplate = func(vm *VM, dcMo *mo.Datacenter, usableDatastores []str
...
+	customSpec := updateCustomSpec(vm, &customSpecItem.Spec)
+
 	cisp := types.VirtualMachineCloneSpec{
 		Location: relocateSpec,
 		Template: false,
 		PowerOn:  false,
 		Config:   &config,
 	}
+	// add the custom spec to vm clone spec if not nil
+	if customSpec != nil {
```
Comment (ashpuri) [inline virtualmachine/vsphere/util.go]: setting this unconditionally will not work?

File: virtualmachine/vsphere/util.go [hunk 5]
```diff
@@ -1229,3 +1259,39 @@ func init() {
 		return nil, NewErrorObjectNotFound(errors.New("Could not find the mob"), name)
 	}
 }
+
+// createCustomSpecStaticIp: creates custom spec for static ip from xml
+func createCustomSpecStaticIp(vm *VM) error {
+	csMgr := object.NewCustomizationSpecManager(
+		vm.client.Client)
```
Comment (ashpuri) [inline virtualmachine/vsphere/util.go]: can be combined with above line

File: virtualmachine/vsphere/constants.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+      </dnsServerList>
+      <dnsSuffixList>
+        <_length>1</_length>
+        <_type>string[]</_type>
+        <e id="0">google.com</e>
+      </dnsSuffixList>
+    </globalIPSettings>
+    <identity>
+      <_type>vim.vm.customization.LinuxPrep</_type>
+      <domain>gsintlab.com</domain>
```
Comment (pankajtandon12) [inline virtualmachine/vsphere/constants.go:32]: Does all these hardcoded xml values gets updated correctly (for an entirely different vcenter setup and configuration) ?

**Review by pankajtandon12 — CHANGES_REQUESTED** [review]

File: virtualmachine/vsphere/constants.go (hunk 6 above)
Comment (avnish30jn) [reply virtualmachine/vsphere/constants.go:32]: Yes, tried it on dev-vcenter and starling-vcenter. The custom spec is created and vm is assigned to the vm.

**Review by pankajtandon12 — APPROVED** [review]
LGTM

**Pull request merged**